includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -1004,16 +1005,18 @@ public function process_payment_for_order( $cart, $payment_information, $additio
 					];
 				} else {
 					$response = [
-						'result'   => 'success',
+						'result'         => 'success',
 						// Include a new nonce for update_order_status to ensure the update order
 						// status call works when a guest user creates an account during checkout.
-						'redirect' => sprintf(
+						'redirect'       => sprintf(
 							'#wcpay-confirm-%s:%s:%s:%s',
 							$payment_needed ? 'pi' : 'si',
 							$order_id,
 							$client_secret,
 							wp_create_nonce( 'wcpay_update_order_status_nonce' )
 						),
+						// Include the payment method ID so the Blocks integration can save cards.
+						'payment_method' => $payment_information->get_payment_method(),
```

Comment:
I'm not sure if this is the best way to get the payment method ID to the Blocks integration. Unfortunately the Blocks integration doesn't provide a good way to forward data from `onPaymentProcessing()` to `onCheckoutCompletedWithSuccess()` without returning something from the server.

An alternative approach to this would be to store the payment method ID through `useState()` and passing a `setPaymentMethod()` function around to update the value. However, this approach seemed simpler; by returning the `payment_method` parameter here we give the `onCheckoutCompletedWithSuccess()` handler exactly what it needs to properly confirm the payment intent in such a way that the card is saved:

https://github.com/Automattic/woocommerce-payments/blob/0883364b9ef3caae97dc3dbd84ee129f4d384d14/client/checkout/blocks/confirm-card-payment.js#L16-L22

I'd be fine with switching to the `useState()` approach though; it'd just be a pure react implementation of this, instead of having this data flow through the checkout process.

Reply:
I like the current approach, we are already sending the `redirect` from the server